src/shacl/ShaclParser.ts

```diff
@@ -0,0 +1,583 @@
+import { TurtleParser } from '../turtle/TurtleParser';
+import {
+  Lexer,
+  IToken,
+  IRecognitionException,
+  IParserConfig,
+  Parser,
+} from 'chevrotain';
+import { getShaclVisitor } from 'shacl/visitor';
+const {
+  getShaclTokenTypes,
+  getShaclTokenMap,
+  categoryTokenMap,
+} = require('./tokens');
+const { turtleTokenMap } = require('../turtle/tokens');
+
+// A SHACL parser for the Turtle serialization of SHACL only. The parser can
+// can accept any arbitrary namespace prefix for SHACL/XSD and still tokenize
+// and parse the document correctly (it will also, of course, parse docuemnts
+// using the full SHACL/XSD IRIs). The parser runs both a parse phase and a
+// second validation phase (using a visitor) in order to accommodate SHACL
+// rules that are not purely syntactic.
+export class ShaclParser extends TurtleParser {
+  private prefixes;
+  private shaclTokenMap;
+  private shaclVisitor: ReturnType<typeof getShaclVisitor>;
+  protected lexer: Lexer;
+
+  // Some SHACL rules cannot be checked for violations during the first parse.
+  // The visitor accepts the CST that results from parsing and checks
+  // conformity with these SHACL rules.
+  private validateWithVisitor = (cst: any) => {
+    // To save resources while parsing, the shaclVisitor is a singleton.
+    if (!this.shaclVisitor) {
+      const BaseSrsVisitor = this.getBaseCstVisitorConstructorWithDefaults();
+      this.shaclVisitor = getShaclVisitor(BaseSrsVisitor);
+    } else {
+      this.shaclVisitor.$resetState();
+    }
+
+    this.shaclVisitor.visit(cst);
+    return this.shaclVisitor.$validateShapes(this.prefixes);
+  };
+
+  public tokenize = (document: string): IToken[] =>
+    this.lexer.tokenize(document).tokens;
+
+  public parse = (
+    document: string
+  ): {
+    errors: IRecognitionException[];
+    semanticErrors: IRecognitionException[];
+    cst: any;
+  } => {
+    this.input = this.tokenize(document);
+    const cst = this.turtleDoc();
+    const { validationErrors } = this.validateWithVisitor(cst);
+    // Next two items are copied so that they can be returned/held after parse
+    // state is cleared.
+    const errors: IRecognitionException[] = [...this.errors];
+    const semanticErrors: IRecognitionException[] = [
+      ...this.semanticErrors,
+      ...validationErrors,
+    ];
+
+    return {
+      errors,
+      semanticErrors,
+      cst,
+    };
+  };
+
+  constructor(
+    config?: Partial<IParserConfig>,
+    prefixes = { shacl: 'sh', xsd: 'xsd' }
+  ) {
+    super(
+      {
+        outputCst: true,
+        recoveryEnabled: true,
+        ...config,
+      },
+      getShaclTokenTypes(prefixes),
```

Comment:
and calling `getShaclTokenMap` 4 times as it's called within `getShaclTokenTypes` as well?

Reply:
i guess you are memoizing it, but..is that necessary for something else or could you just `const` the call to `getShackTokenTypes` then pass it in as an argument to `getShackTokenMap`?

Reply:
You can't create a `const` before the `super` call, so we have to call `getShaclTokenTypes` at least 3 times (twice inside the initial `super` call, then one more time after that to pass the tokens to the lexer). This is (part of) why I memoized it, so that it's basically a no-op after the first call. `getShaclTokenTypes` needs the token map, and we can't `const` the result of `getShaclTokenMap` it before the call to `super` either, so either way we're also going to need to call _it_ multiple times. Since there's also almost no difference between calling the memoized versions more than once (since, after the first call, they just return the previous result) and creating a separate `const`, going any further here didn't seem worth it.

Reply:
basically b/c JS? https://stackoverflow.com/questions/1168345/why-do-this-and-super-have-to-be-the-first-statement-in-a-constructor

Reply:
Yep, pretty much.

Reply:
👎

Reply:
Trust me, I wanted to do exactly what you were thinking. :)

Most of the other parsers don't have this problem because the tokens are static (they are keywords/don't depend on prefixes). SHACL is a special case because the 'sh:' alias could be changed to anything, and the parser accounts for that. Maybe this was some over-engineering (we could just support only the standard alias) but it at least affords some flexibility that is technically allowed by the spec.